File: 113-path-sum-ii/113.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root, target):
        if root is None:
            return

        self.sum += root.val
        self.sets.append(root.val)

        self.dfs(root.left, target)
        self.dfs(root.right, target)

        if self.sum == target:
            if not root.left and not root.right:
                self.ans.append(self.sets.copy())
        self.sum -= root.val
        self.sets.pop()

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        self.sum = 0
        self.ans = []
        self.sets = []
        self.dfs(root, targetSum)

        return self.ans

File: 113-path-sum-ii/113.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<vector<int>> ans;
    vector<int> tmp;
    void bt(TreeNode *node, int targetSum, int currentSum) {
        if (!node)
            return;
        currentSum += node->val;
        tmp.push_back(node->val);
        if (currentSum == targetSum && !node->left && !node->right) {
            ans.push_back(tmp);
        } else {
            bt(node->left, targetSum, currentSum);
            bt(node->right, targetSum, currentSum);
        }
        tmp.pop_back();
    }
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        bt(root, targetSum, 0);
        return ans;
    }
};
